library/management/chat/src/main/java/org/quiltmc/qsl/chat/api/types/AbstractChatMessage.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2023 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.chat.api.types;
+
+import net.minecraft.entity.player.PlayerEntity;
+import org.jetbrains.annotations.NotNull;
+import org.quiltmc.qsl.chat.api.QuiltMessageType;
+
+import java.util.EnumSet;
+
+/**
+ * An abstract message, extended for the various types of messages.
+ * @param <S> The type of the serialized form of this class, usually a Packet<?> but not required
+ */
+public abstract class AbstractChatMessage<S> {
+	protected final @NotNull PlayerEntity player;
+	protected final boolean isOnClientSide;
+
+	protected AbstractChatMessage(@NotNull PlayerEntity player, boolean isOnClientSide) {
+		this.player = player;
+		this.isOnClientSide = isOnClientSide;
+	}
+
+	// These are mostly annotated as @NotNull for dev productivity, I get nice warnings about
+	// What parts of the subclasses I need to implement still
+
+	/**
+	 * @return the set of {@link QuiltMessageType}s associated with this message
+	 */
+	public abstract @NotNull EnumSet<QuiltMessageType> getTypes();
+
+	/**
+	 * @return this message in its serialized form
+	 */
+	public abstract @NotNull S serialized();
+
+	/**
+	 * Returns the player associated with this packet, which changes meaning based on the sidedness and direction.
+	 * <p>
+	 * For System and Profile Independent messages, the player is the target of the message as players cannot send these to the server.
+	 * <p>
+	 * For Chat messages, on the server inbound, it is the sender of the chat message, while outbound it is the player the message will be sent to.
+	 * Chat messages on the client return the client player.
+	 * @return the {@link PlayerEntity} associated with this packet
+	 */
+	public final @NotNull PlayerEntity getPlayer() {
+		return player;
+	}
+
+	/**
+	 * @return if this message is on the client side or not. This can also be determined through checking for {@link QuiltMessageType#CLIENT}
```

Comment:
Hm, I think the second sentence would be split off the return? wait oh no this isn't a `{@return}`; Yeah, maybe something better would be:
```suggestion
	 * {@return if this message is on the client side or not}
	 * <p>
	 * This can also be determined through checking for {@link QuiltMessageType#CLIENT}.
```

Reply:
![image](https://user-images.githubusercontent.com/52360088/212801645-6a92d026-b9c1-4917-b849-89c4ffa44525.png)
IDEA *realllly* doesnt like that pattern, removing the `{}` seems to work though?
![image](https://user-images.githubusercontent.com/52360088/212801722-39314e2d-25f8-4f0e-8d43-86575a4a6ac5.png)

Reply:
wait, it did that all along?
Yeah, the intention is to deduplicate information, although yeah, as an IntelliJ newbie, the raw `@return` does not seem good;
@LambdAurora is this supposed to happen?